### PR TITLE
Support ?Trait bounds in supertraits and dyn Trait under a feature gate

### PR DESCRIPTION
**Pull request by Bryanskiy**
This patch allows `maybe` polarity bounds under a feature gate. The only language change here is that corresponding hard errors are replaced by feature gates. Example: 
```rust
#![feature(more_maybe_bounds)]
...
trait Trait1 : ?Trait { ... } // ok
fn foo(_: Box<(dyn Trait2 + ?Trait)>) {} // ok
fn bar<T: ?Sized + ?Trait>(_: &T) {} // ok
```
Maybe bounds still don't do anything (except for `Sized` trait), however this patch will allow us to [experiment with default auto traits](https://github.com/rust-lang/rust/pull/120706#issuecomment-1934006762).

This is a part of the [MCP: Low level components for async drop](https://github.com/rust-lang/compiler-team/issues/727)

### CODE REVIEW

**Comment by Nilstrieb** [comment]
for reference, is there any language team shepherd for all of this default bound experimentation? I only ever see random work and don't remember seeing design documents or anything like that

File: compiler/rustc_error_codes/src/error_codes/E0203.md
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0203.md:1]: :+1:

File: compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:642]: Any specific reason why `?Trait` bounds are skipped here?

**Comment by petrochenkov** [comment]
Need to link to https://github.com/rust-lang/compiler-team/issues/727 from this PR description.
@rustbot author

**Review by petrochenkov — APPROVED** [review]

**Comment by petrochenkov** [comment]
Implementation LGTM, except the commit message needs to be more human oriented.

I'm going to pass this to r? @traviscross in cases some additional lang team approval is needed, but otherwise r=me.

**Comment by Bryanskiy** [comment]
In the last update `MisplacedRelaxTraitBound` was replaced by feature gate.

**Comment by traviscross** [comment]
@rustbot labels +T-lang +I-lang-nominated

Since there's no accepted RFC, note that this would need to go in under an *experimental* feature flag.  See the description of this here:

https://lang-team.rust-lang.org/how_to/experiment.html

We'll need to discuss this in a lang meeting for visibility and to ensure this experiment has been properly approved, so let's nominate this.

**Comment by petrochenkov** [comment]
Summary:
- https://github.com/rust-lang/rust/pull/120706 introduces a way to add new auto traits that are appended to all bound lists by default, similarly to existing `Sized`. Such traits may include `Leak`, `SyncDrop` or similar, see https://github.com/rust-lang/rust/pull/120706#issuecomment-1934006762 for more detailed motivation.
- To opt out from bounds added by default the `?Trait` syntax is used, but such "maybe" bounds are not supported in some contexts like supertrait lists and `dyn Trait + ...` lists, because `Sized` is not added by default in those context.
- This PR adds a feature for supporting `trait Trait1: ?Trait2`, `dyn Trait1 + ?Trait2` and also multiple maybe bounds in the same list `?Trait1 + ?Trait2`, because the new traits need to be added by default in those contexts too, and `?Sized + ?Leak` may also make sense.
- We need this to be available in bootstrap compiler, to make experiments on standard library without adding too many `#[cfg(not(bootstrap))]`s

**Comment by Bryanskiy** [comment]
In the last update https://github.com/rust-lang/rust/pull/120706#discussion_r1524754002 was fixed.

**Comment by Dylan-DPC** [comment]
@Bryanskiy what's the status of this? thanks

**Comment by Bryanskiy** [comment]
@Dylan-DPC PR blocked on https://github.com/rust-lang/rust/pull/120706

**Comment by Bryanskiy** [comment]
@petrochenkov https://github.com/rust-lang/rust/pull/120706 was [accepted](https://github.com/rust-lang/rust/pull/120706#issuecomment-2248093357) for experiment. Let's merge this first.

File: tests/ui/maybe-bounds.stderr
Comment (Bryanskiy) [inline tests/ui/maybe-bounds.stderr:9]: Should such diagnostic changes be visible only on nightly?

Comment (petrochenkov) [reply tests/ui/maybe-bounds.stderr:9]: It's already only visible on nightly according to the logic in `fn add_feature_diagnostics_for_issue`.

**Comment by matthiaskrgr** [comment]
@bors r-
tests need rebase/blessing I guess https://github.com/rust-lang/rust/pull/128191#issuecomment-2250972034

**Comment by Bryanskiy** [comment]
The latest commit fixes [this](https://github.com/rust-lang/rust/pull/120706#discussion_r1691802121)

@rustbot ready

**Pull request merged**
